docs/storage/commit_log/commit_log_based_storage_design.md

```diff
@@ -0,0 +1,361 @@
+
+#### Master election
+
+Multiple sequencers may be running to provide resilience, if this is the case
+there must be a mechanism for choosing a single master instance among the
+running sequencers. The Trillian repo provides an etcd-backed implementation
+of this already.
+
+A sequencer must only participate/remain the master if its local database state
+is at least as new at the latest message in the Kafka `STHs` topic.
+
+The current master sequencer will create new STHs and publish them to the
+`STHs` topic, the remaining sequencers will run in a "mirror" mode to keep
+their local database state up-to-date with the master.
+
+#### Local DB storage
+
+This does not *need* to be transactional, because writes should be idempotent,
+but the implementation of the Trillian storage driver must buffer *all*
+writes and only attempt to apply them to the local storage when `Commit` is
+called.
+
+The write of an updated STH to local storage needs slightly special attention,
+in that it must be the last thing written by `Commit`, and must only be written
+if all other buffered writes succeeded.
+
+In the case of a partial commit failure, or crash of the signer, the next
+sequencing cycle should find that identical writes are re-attempted due to the
+signer process outlined below.
+
+#### Sequencing
+
+Assigning sequence numbers to queued leaves is implicitly performed by the
+addition of entries to the Kafka `Leaves` topic (this is termed *offset* in
+Kafka documentation).
+
+
+##### Abstract Signer process
+
+```golang
+func SignerRun() {
+  // if any of the below operations fail, just bail and retry
+
+  // read `dbSTH` (containing `treeRevision` and `sthOffset`) from local DB
+  dbSTH.treeRevision, dbSTH.sthOffset = tx.LatestSTH()
+
+  // Sanity check that the STH table has what we already know.
+  ourSTH := kafka.Read("STHs/<treeID>", dbSTH.sthOffset)
+  if ourSTH == nil {
+    glog.Errorf("should not happen - local DB has data ahead of STHs topic")
+    return
+  }
+  if ourSTH.expectedOffset != dbSTH.sthOffset {
+    glog.Errorf("should not happen - local DB committed to invalid STH from topic")
+    return
+  }
+  if ourSTH.timestamp != dbSTH.timestamp || ourSTH.tree_size != dbSTH.tree_size {
+    glog.Errorf("should not happen - local DB has different data than STHs topic")
+    return
+  }
+
+  // Look to see if anyone else has already stored data just ahead of our STH.
+  nextOffset := dbSTH.stdOffset
+  nextSTH := nil
+  for {
+      nextOffset++
+      nextSTH = kafka.Read("STHs/<treeID>", nextOffset)
+      if nextSTH == nil {
+        break
+      }
+      if nextSTH.expectedOffset != nextOffset {
+        // Someone's been writing STHs when they weren't supposed to be, skip
+        // this one until we find another which is in-sync.
+        glog.Warning("skipping unexpected STH")
+        continue
+      }
+      if nextSTH.timestamp < ourSTH.timestamp || nextSTH.tree_size < ourSTH.tree_size {
```

Comment:
I think the loop at L212 was a bit of old junk that somehow made its way in here, I've removed it and fixed up some other stuff which seemed bogus.